### ISSUE TITLE
Jogamp 2.3.2 profile detection fails on Mesa

### ISSUE THREAD

**Issue by nmanpkr** [runelite/runelite#11357]
Under Ubuntu 18.04 - Ubuntu 20.04 and possibly other distributions of Linux, GPU plugin causes client crash when switched on using .Jar and causes Thrown Exception when using AppImage. The issue is caused by Jogamp 2.3.2, Jogamp fails to detect OpenGL profile since Gallium was dropped from renderer string using mesa minimum required version up to Mesa 20.0.0.4. This issue is fixed in the dev builds of Jogamp but not the stable.

To reproduce:
1) Start client
2) Open Settings
3) Enable GPU plugin (Either from login, or in-game no difference)

Setup:
OS: Ubuntu 18.04, 19.10, and Ubuntu 20.04
Hardware: Radeon HD 7660D
Client is ran with Hardware Acceleration Disabled

Error: (AppImage .Jar crashes with no crash logs)
2020-04-23 22:23:20 [Client] ERROR n.r.client.plugins.gpu.GpuPlugin - Error starting GPU plugin
java.lang.InternalError: XXX0 profile[1]: GL3bc -> profileImpl GL4bc !!! not mapped

**Comment by nmanpkr** [comment]
Figured I would post an update on where I'm at with this. I'm trying my best to tackle this before going and buying an Nvidia card and using proprietary drivers. So here is what I have tried.

I have a "patched" version of GLContextImpl inside Jogamp.jar from Jogl forums. Basically returns true for isMesa boolean for both renderer strings Gallium and Mesa.

It is not as simple as placing the patched jog-all-2.3.2.jar inside .runelite/repository2/ because the launcher checks the hashes against each library and marks them for removal and replacement on launch. To get around this I decompiled the Launcher class and added two conditional statements to skip adding jog-all-2.3.2.jar to the artifacts download list, and to also skip the hash checksum.

if(artifact.getName().equals("jog-all-2.3.2.jar"))
continue;

Unfortunately the client still instantly crashes without a log. I would try this with the AppImage to see if the error message changes, but I'm not sure how the AppImage is packed.

I tried making jog-all-2.3.2.jar immutable using chattr -i jog-all-2.3.2.jar and running AppImage with the patched Jogl, but the client throws and exception when the write permission fails so this was a no go.

There are two ideas I have not tried:
I have not tried to build jog-all from source and setting isMesa to true without a check, and using this with my modified Launcher.

The other idea I have not tried, which is more of a way to replace jog-all-2.3.2 without tearing into the launcher. The workaround in mind is hosting a local webserver on client launch with a modified bootstrap.json that points to a locally hosted jog-all-2.3.2.jar with an updated file size and hash. I feel this is more of a last resort if I can even get the plugin to work using the above methods.

I'm at a loss here and really am not sure where to go from here.

**Comment by abextm** [comment]
Here is a build of RuneLite with a newer version of jogl: 
https://abex.sauville.org/client-1.6.13-SNAPSHOT-shaded-joglnew.jar

Can you test this and see if it fixes your issue?

Also you don't have to decompile (or build) the launcher, you can just build the client is this repo and run the shaded jar in `runelite-client/target/`

**Comment by nmanpkr** [comment]
> Here is a build of RuneLite with a newer version of jogl:
> https://abex.sauville.org/client-1.6.13-SNAPSHOT-shaded-joglnew.jar
> 
> Can you test this and see if it fixes your issue? You need to run it with java>=11
> 
> Also you don't have to decompile (or build) the launcher, you can just build the client is this repo and run the shaded jar in `runelite-client/target/`

Running this client returns a similar error:
java.lang.InternalError: XXX0 profile[2]: GL2 -> profileImpl GL3bc !!! not mapped

I think it might be useful to supply my hardware information:

glxinfo | grep OpenGL

OpenGL vendor string: X.Org
OpenGL renderer string: AMD ARUBA (DRM 2.50.0 / 5.4.0-26-generic, LLVM 9.0.1)
OpenGL core profile version string: 4.3 (Core Profile) Mesa 20.0.4
OpenGL core profile shading language version string: 4.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.1 Mesa 20.0.4
OpenGL shading language version string: 1.40
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.1 Mesa 20.0.4
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
OpenGL ES profile extensions:

In addition, I have ran the master-branch RuneLite client with the parameters printed to terminal when launching with the launcher with my "patched" version of jogl and the current linux-natives and the client didn't crash when turning on the GPU plugin. Instead the plugin simply switched off and printed out an error: Inconsistency detected by ld.so: dl-lookup.c: 111: check_match

I resolved this by installing openJDK8 alongside JDK11 and launching using java 8 and the plugin switches on and the client doesn't crash, however the client freezes and prints out: Profile GL3bc is not available on X11GraphicsDevice which from my quick research points to a problem using the newer natives with openJDK8. 

tldr; In conclustion, I believe JOGL is at fault and this won't be resolved until they release 2.4.0 which states a bug fix for: GL3bc -> profileImpl GL4bc !!! not mapped, however this has been proposed for a long time. So at this point I think I am out of luck.

**Comment by ZaneA** [comment]
I ran into this after upgrading to Ubuntu 20.04, I previously had the GPU plugin enabled and couldn't find a way to start the client at all after the upgrade.

```
$ glxinfo | grep OpenGL
OpenGL vendor string: Intel
OpenGL renderer string: Mesa Intel(R) HD Graphics 620 (KBL GT2)
OpenGL core profile version string: 4.6 (Core Profile) Mesa 20.0.4
OpenGL core profile shading language version string: 4.60
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 4.6 (Compatibility Profile) Mesa 20.0.4
OpenGL shading language version string: 4.60
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 20.0.4
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
OpenGL ES profile extensions:
```

> Here is a build of RuneLite with a newer version of jogl:
> https://abex.sauville.org/client-1.6.13-SNAPSHOT-shaded-joglnew.jar
> 
> Can you test this and see if it fixes your issue? You need to run it with java>=11
> 
> Also you don't have to decompile (or build) the launcher, you can just build the client is this repo and run the shaded jar in `runelite-client/target/`

Thanks! This build worked for me on OpenJDK 11 :)

```
$ java -version                                                                                           
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Ubuntu-3ubuntu1)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Ubuntu-3ubuntu1, mixed mode, sharing)
```

Another workaround if the client isn't launching at all, you can use `export LIBGL_ALWAYS_SOFTWARE=1` before running, not ideal but good enough in the mean time.

**Comment by Adam-** [comment]
> 
> 
> I ran into this after upgrading to Ubuntu 20.04, I previously had the GPU plugin enabled and couldn't find a way to start the client at all after the upgrade.
> 
> ```
> $ glxinfo | grep OpenGL
> OpenGL vendor string: Intel
> OpenGL renderer string: Mesa Intel(R) HD Graphics 620 (KBL GT2)
> OpenGL core profile version string: 4.6 (Core Profile) Mesa 20.0.4
> OpenGL core profile shading language version string: 4.60
> OpenGL core profile context flags: (none)
> OpenGL core profile profile mask: core profile
> OpenGL core profile extensions:
> OpenGL version string: 4.6 (Compatibility Profile) Mesa 20.0.4
> OpenGL shading language version string: 4.60
> OpenGL context flags: (none)
> OpenGL profile mask: compatibility profile
> OpenGL extensions:
> OpenGL ES profile version string: OpenGL ES 3.2 Mesa 20.0.4
> OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
> OpenGL ES profile extensions:
> ```
> 
> > Here is a build of RuneLite with a newer version of jogl:
> > https://abex.sauville.org/client-1.6.13-SNAPSHOT-shaded-joglnew.jar
> > Can you test this and see if it fixes your issue? You need to run it with java>=11
> > Also you don't have to decompile (or build) the launcher, you can just build the client is this repo and run the shaded jar in `runelite-client/target/`
> 
> Thanks! This build worked for me on OpenJDK 11 :)
> 
> ```
> $ java -version                                                                                           
> openjdk version "11.0.7" 2020-04-14
> OpenJDK Runtime Environment (build 11.0.7+10-post-Ubuntu-3ubuntu1)
> OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Ubuntu-3ubuntu1, mixed mode, sharing)
> ```
> 
> Another workaround if the client isn't launching at all, you can use `export LIBGL_ALWAYS_SOFTWARE=1` before running, not ideal but good enough in the mean time.

Does this build work? https://github.com/runelite/runelite/issues/6482#issuecomment-621504997 It is from the OSX work I am doing and necessarily includes a jogl update aswell.

**Comment by ZaneA** [comment]
@Adam- yup! that build is working fine for me as well :)

**Comment by Adam-** [comment]
fixed in f95c5af4e4acb3dea94700bf5e305a5a22a07b1c
